### ISSUE TITLE
Scylla fails to synchronize materialized view in batch statement when more than 19 materialized views created based on particular table

### ISSUE THREAD

**Issue by Jager227** [scylladb/scylladb#17469]
Creating keyspace, table and views:

```
CREATE KEYSPACE andrey_test WITH replication = {'class': 'SimpleStrategy', 'replication_factor': '1'}  AND durable_writes = true;

CREATE TABLE andrey_test.values (
    type text,
    uid text,
    field text,
    value text,
    change timestamp,
    PRIMARY KEY ((type, uid), field, value)
);
CREATE MATERIALIZED VIEW andrey_test.ix_test00 AS
    SELECT *
    FROM andrey_test.values
    WHERE type IS NOT null AND uid IS NOT null AND field = 'test00' AND value IS NOT null
    PRIMARY KEY ((type, field, value), uid);
CREATE MATERIALIZED VIEW andrey_test.ix_test01 AS
    SELECT *
    FROM andrey_test.values
    WHERE type IS NOT null AND uid IS NOT null AND field = 'test01' AND value IS NOT null
    PRIMARY KEY ((type, field, value), uid);
```
Creating views up to andrey_test.ix_test19 with the same logic.

--------------------------------------------------------------------------------------
Use this sequence of commands to reproduce the bug:
```
INSERT INTO andrey_test.values (type,uid,field,value,change) VALUES ('type0','uid0','test19','forchange',toTimestamp(now()));
BEGIN BATCH

  INSERT INTO andrey_test.values (type,uid,field,value,change) VALUES ('type0','uid0','field0','0',toTimestamp(now()));

  INSERT INTO andrey_test.values (type,uid,field,value,change) VALUES ('type0','uid0','field1','1',toTimestamp(now()));

  DELETE FROM andrey_test.values WHERE type='type0' AND uid='uid0' AND field='field1' AND value='0';

  INSERT INTO andrey_test.values (type,uid,field,value,change) VALUES ('type0','uid0','field2','2',toTimestamp(now()));

  DELETE FROM andrey_test.values WHERE type='type0' AND uid='uid0' AND field='field2' AND value='1';

  INSERT INTO andrey_test.values (type,uid,field,value,change) VALUES ('type0','uid0','test19','changed',toTimestamp(now()));

  DELETE FROM andrey_test.values WHERE type='type0' AND uid='uid0' AND field='test19' AND value='forchange';

APPLY BATCH;

SELECT * FROM andrey_test.values WHERE type='type0' AND uid='uid0' AND field='test19';

SELECT * FROM andrey_test.ix_test19 WHERE type='type0' AND field='test19' AND value IN ('forchange','changed');
```
-------------------------------------------------------------
Results:
```
cqlsh> SELECT * FROM andrey_test.values WHERE type='type0' AND uid='uid0' AND field='test19';

 type  | uid  | field  | value   | change
-------+------+--------+---------+---------------------------------
 type0 | uid0 | test19 | changed | 2024-02-22 08:17:52.993000+0000

(1 rows)
cqlsh> SELECT * FROM andrey_test.ix_test19 WHERE type='type0' AND field='test19' AND value IN ('forchange','changed');

 type  | field  | value     | uid  | change
-------+--------+-----------+------+---------------------------------
 type0 | test19 |   changed | uid0 | 2024-02-22 08:17:52.993000+0000
 type0 | test19 | forchange | uid0 | 2024-02-22 08:17:52.923000+0000
(2 rows)
```

As we can see, line with value 'forchange' was deleted from base table but remained in materialized view. 
This bug manifest itself only when there are more than 19 materialized views created to the base table.

**Comment by Jager227** [comment]
> Which version of Scylla are you using?
Version 5.5.0

**Comment by nyh** [comment]
Wow, just two weeks ago I fixed what seems a very similar bug - see https://github.com/scylladb/scylladb/pull/17164. In that other bug, if we sent a large batch (in that case, actually a "IN") of clustering-prefix deletions, but one out of every 25 view updates was lost on the view. It turned out we worked internally in batches of 100 view-update steps, and because of silly reasons each range deletion was counted as 4 steps, and then we had a bug where we could lose an update across the 100-step boundary.

I wonder if **this** issue isn't the same - you do a batch of updates multiplied by 19 views and the result might be more than 100 view updates and encounter the same bug.

I'll prepare a test for your scenario and try to reproduce it, but can you please do me a favor and test your own reproducer on Scylla's nightly docker, which includes https://github.com/scylladb/scylladb/pull/17164? That fix hasn't yet reached any official version, and I want to check maybe it fixes your bug too.

**Comment by Jager227** [comment]
Alright, it seems that your patch fixed the issue. Should have checked it on a fresh version... 

Should I close this issue?

**Comment by nyh** [comment]
> @nyh Alright, it seems that your patch fixed the issue. Should have checked it on a fresh version...

Very nice, so closing as a duplicate of https://github.com/scylladb/scylladb/issues/17117. It's good that it's not yet another bug, but it's sad that that bug I already fixed if even more severe than I thought (and not limited to range tombstones like in our previous reproducers).

I'll go backport the fix to 5.4 and 5.2 now.

I wanted to create a new regression test based on your example, but it's hard to tell what is important and what is not. I'll try to spend a bit more time on it but not sure I'll get around to it.

Thanks for reporting the issue! I shows how important it is to backport the fix...

**Comment by Jager227** [comment]
> > > Which version of Scylla are you using?
> > 
> > 
> > Version 5.5.0
> 
> This is the master - and the fix is already in master - so how come you were seeing this?

The fix is recent, just didn't update my local branch for a while
